Fix #1348: Rework the Record Encryption documentation describing the role of the administrator (#1413)
kroxylicious/kroxylicious: docs/modules/record-encryption/aws-kms/con-aws-kms-setup.adoc

will be to prefix the alias used by the filter with the word `KEK_`. If a different naming convention is used, adapt

> the instructions accordingly.

Reviewer: Suggestion for lines 10-13:

> Establish a naming convention for key aliases to keep the filter’s keys separate from those used by other systems. 
> Here, we use a prefix of `KEK_` for filter aliases. 
> Adjust the instructions if a different naming convention is used.

Author: Done

== Role of the Administrator

In order to use the filter, there must be an external entity which is responsible for creating encryption keys within AWS KMS which will be used to encrypt the records. It is up to the organization

> deploying Record Encryption to decide how this entity is managed.

Author: The reason we chose this language was because the "admin" could be a human being, or could be another system.  We are trying to convey "something outside" of Kroxylicious.  Do you have a suggestion?

> deploying Record Encryption to decide how this entity is managed.

Reviewer 2: "administrative process" perhaps? that leaves it ambiguous as to whether it's an automated or manual process...

> deploying Record Encryption to decide how this entity is managed.

Author: how about introducing with "administrator (or administrative process)" and then later references use "administrator". @PaulRMellor @SamBarker WDYT?

> deploying Record Encryption to decide how this entity is managed.

Reviewer 3: Yeah I think that works

> deploying Record Encryption to decide how this entity is managed.

Reviewer: That's great Keith

> deploying Record Encryption to decide how this entity is managed.

Author: @PaulRMellor pushed up a change, how does that look?

> deploying Record Encryption to decide how this entity is managed.

Reviewer: Looks good. As mentioned, a few unexpanded comments were mised.

The *Administrator* must have permissions to create keys in AWS KMS.  As a starting point, the
built-in AWS policy `AWSKeyManagementServicePowerUser` confers sufficient key management privileges.

To help get you started, the following commands can be used to establish an Administrator user with permissions
suitable for managing encryption keys in KMS via an AWS Cloud Shell. This example illustrates using the user-name

> `kroxylicious-admin` but the choice of name is not significant.  If a different user-name is used, adapt the

Reviewer 4: Didn't we agreed to add a comment here for testing purposes to fulfill this issue #1348 ?? For those cases where a deletion of the key is needed.

I think adding a link to the AWS developer docs would be sufficient.

Author: On #1348 i commented

> I don't think there is a case for documenting the permissions required for deleting keys until there exists a procedure to work out if a key is not in use.

so I think we should not add the link right now.

will be to prefix the alias used by the filter with the word `KEK_`. If a different naming convention is used, adapt
the instructions accordingly.

> == Role of the Administrator

Author: I thought about alternative terms: "Key Admin", "KEK Admin", "Key Manager".. in the end I went back to Administrator.

Reviewer 3: My only other suggestion would be `The Administrative Actor` 

but I think the `Role of the Administrator` is probably a better title.

> In order to use the filter, there must be an *Administrator*.  The Administrator has the responsibility
> to create encryption keys within AWS KMS which will be used to encrypt the records. It is up to the organization
> deploying Record Encryption to decide how the role of the *Administrator* maps into their organization structure.

Reviewer 3: ```suggestion
In order to use the filter, there must be an external entity which is responsible for creating encryption keys within AWS KMS which will be used to encrypt the records. It is up to the organization
deploying Record Encryption to decide how this entity is managed.
```

I worry a bit that talking about an Administrator and mapping to the org structure assumes a human rather than some cloud formation or terraform code to manage it via git ops. 

That said talking about Administrators and org structures maybe makes our point more intelligible.

> The *Administrator* must have permissions to log in to AWS and create keys in AWS KMS.  As a starting point, the

Reviewer 3: I'd say `log in to AWS` is implicit within `create keys in AWS KMS` without pre supposing a human interaction.

built-in AWS policy `AWSKeyManagementServicePowerUser` confers sufficient key management privileges.

To help get you started, the following commands can be used to establish an Administrator user with permissions
suitable for managing encryption keys in KMS via an AWS Cloud Shell. This example illustrates using the user-name

> `kroxylicious-admin` but the choice of name is not significant.  If a different user-name is used, adapt the

Reviewer 3: perfect.

echo Now log in at ${CONSOLE_URL}  with user name ${ADMIN} password ${INITIAL_PASSWORD} and change the password.
----

> == Filter Actor

Author: I got rid of the term 'Actor'.   I'm now referring to the entity within the KMS (Token in the case of Vault, IAM User in the case of AWS).

> == Filter IAM Identity

Reviewer 3: I don't know what `Filter IAM Identity` means. 

I think just using `IAM Identity` as the heading is clear enough.

Author: "IAM Identity" alone wouldn't get across that this the identity that will be used by the Filter. How about 'IAM Identity for the Filter'?

Author: I used the title "Establish an application identity for the Filter"

The filter needs to be able to authenticate to AWS in order to perform the operations needed for Envelope Encryption
(generating and decrypting DEKs).  To this end, there must be an AWS IAM identity established for the filter itself with
sufficient permissions.

Use AWS IAM to create the `kroxylicious` user. Create an *Access Key* for this user. The Access Key/Secret Key pair
will be used by the Filter. Do not enable the Console for this user.

> If using the CLI, these commands can be used to establish the Filter IAM Identity.  This example illustrates using the

Reviewer 3: ```suggestion
If using the CLI, these commands can be used to establish the IAM Identity for the Filter.  This example illustrates using the
```
or 
```suggestion
If using the CLI, these commands can be used to establish the Filter's IAM Identity.  This example illustrates using the
```

Its the IAM identity belonging to the filter.

aws iam create-access-key --user-name kroxylicious
----

> == Create Alias Based Policy for keys matching the KEK aliasing convention

Reviewer: ```suggestion
== Create an alias-based policy for KEK aliases
```

aws iam create-policy --policy-name KroxyliciousRecordEncryption --policy-document file:///tmp/policy
----

== Apply Alias Based Policy to Filter IAM Identity

> Attach the alias policy to the Filter IAM Identity.  This will allow the Filter to invoke the

Reviewer 3: Similar thought to above `Filter's IAM Identity`. 

I'm not sure about capitalising the `F` in filter either...
